### PR TITLE
#164691220 Add User profiles Functionality

### PR DESCRIPTION
**Pull request by NicholusMuwonge**
#### What does this PR do?
Enables user to create a profile and also update it.

#### Description of Task to be completed?
**To get the following end points working**
- ```**PUT** http://127.0.0.1:8000/api/user/```
- ```**GET** http://127.0.0.1:8000/api/user/``` 
- ```**GET** http://127.0.0.1:8000/api/profiles/<str:username>```

#### How should this be manually tested?
**One** should be able to clone this repo and test out the code In Postman or any other tool of their choice to test it out.

- Added the relational model which creates a profile on creation of a user.
- User can update their profile if they have the right authorisation using the route ```http://127.0.0.1:8000/api/user/ ``` 
The update body should look like.
```
{
  "user":{
    "username": "nicksbros",
    "email": "some@some.com",
    "bio":"I work at a statefarm i",
    "image": "https://inspector"
    
}
}
```
- User can access their own profile using a get method using routes ```http://127.0.0.1:8000/api/user/``` or ``` http://127.0.0.1:8000/api/profiles/<str:username> ```
response should look like 
```
{
    "user": {
        "email": "some@some.com",
        "username": "nicksbros",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNuIjoibmlja3Nicm9zIiwiZXhwIjoxNTU2OTk2MTkyfQ.W4XnFcVppUCfj6_vhF6TuGjXxoYCzbNDe1EeN6mP4E0",
        "bio": "I work at a statefarm i",
        "image": ""
    }
}
```
the image field shoul be black since it only returns an image object.

- User should be able to view other people's profiles using route ``` http://127.0.0.1:8000/api/profiles/<username> ``` and the response object should look like 

``` {
    "profile": {
        "username": "nicksbrosty",
        "bio": "I am cool bro",
        "image": "https://static.productionready.io/images/smiley-cyrus.jpg"
    }
}
```
with a default route which represents the vatar to act as place holder for user's profile picture. 
- User shouldnt be able to edit one's profile as for the PUT method for this route is blocked. like
```{
    "profile": {
        "detail": "Method \"PUT\" not allowed."
    }
}
```
- If user doesnt exist, response should look like ``` {
    "errors": {
        "detail": "The requested profile does not exist."
    }
}```

- A time stamp for created_at and updated_at

**Story Id** https://www.pivotaltracker.com/story/show/164691220

### CODE REVIEW

File: authors/apps/authentication/renderers.py [hunk 1]
```diff
@@ -1,24 +1,32 @@
...
 
-        if errors is not None:
-            # As mentioned about, we will let the default JSONRenderer handle
-            # rendering errors.
-            return super(UserJSONRenderer, self).render(data)
+        # token = data.get('token', None)
+        # if token is not None and isinstance(token, bytes):
+        #     # Also as mentioned above, we will decode `token` if it is of type
+        #     # bytes.
+        #     data['token'] = token.decode('utf-8')
```
Comment (reifred) [inline authors/apps/authentication/renderers.py]: Kindly get rid of this commented code.

File: authors/apps/authentication/renderers.py [hunk 2]
```diff
@@ -1,24 +1,32 @@
...
 
         # Finally, we can render our data under the "user" namespace.
-        return json.dumps({
-            'user': data
-        })
+        # return json.dumps({
+        #     'user': data
+        # })
+        return super(UserJSONRenderer, self).render(data)
```
Comment (reifred) [inline authors/apps/authentication/renderers.py]: In addition, I also see the default user render was also commented out. I don't know whether you catered for it with existing changes.

File: requirements.txt [hunk 3]
```diff
@@ -19,3 +19,5 @@ requests==2.21.0
 six==1.12.0
 urllib3==1.24.1
 whitenoise==4.1.2
+py3dns==3.2.0
+validate-email==1.3
```
Comment (reifred) [inline requirements.txt]: You could also remove the validate-email, and py3dns modules since you are no longer using them

**Review by reifred — CHANGES_REQUESTED** [review]
Thanks, @NicholusMuwonge for the work done. Kindly address these changes as I test the application.

File: authors/apps/authentication/renderers.py (hunk 2 above)
Comment (NicholusMuwonge) [reply authors/apps/authentication/renderers.py]: fixed please

File: requirements.txt (hunk 3 above)
Comment (NicholusMuwonge) [reply requirements.txt]: fixed please

**Review by richien — CHANGES_REQUESTED** [review]
Hi Nick, it appears that you are trying to merge two branches from this PR since you are merging two commits, one for this feature and another for the signup validation feature. Perhaps you accidentally created this branch off the signup validation branch, instead of from develop. Kindly fix this.

**Comment by NicholusMuwonge** [comment]
> Hi Nick, it appears that you are trying to merge two branches from this PR since you are merging two commits, one for this feature and another for the signup validation feature. Perhaps you accidentally created this branch off the signup validation branch, instead of from develop. Kindly fix this.

Its because, I have two commits that are mine that need to be merged to develop at the same time

**Comment by richien** [comment]
> 
> 
> > Hi Nick, it appears that you are trying to merge two branches from this PR since you are merging two commits, one for this feature and another for the signup validation feature. Perhaps you accidentally created this branch off the signup validation branch, instead of from develop. Kindly fix this.
> 
> Its because, I have two commits that are mine that need to be merged to develop at the same time

That's ok, I understand that you want to merge two commits however the convention is that each PR should merge 1 commit, which is my you have a PR for validating signup. Once the signup validation is approved and merged to develop, then you can rebase this branch with develop to add that feature to this branch. The way you have done it violates that flow and is bound to cause some conflicts down the line.

**Comment by NicholusMuwonge** [comment]
> > > Hi Nick, it appears that you are trying to merge two branches from this PR since you are merging two commits, one for this feature and another for the signup validation feature. Perhaps you accidentally created this branch off the signup validation branch, instead of from develop. Kindly fix this.
> > 
> > 
> > Its because, I have two commits that are mine that need to be merged to develop at the same time
> 
> That's ok, I understand that you want to merge two commits however the convention is that each PR should merge 1 commit, which is my you have a PR for validating signup. Once the signup validation is approved and merged to develop, then you can rebase this branch with develop to add that feature to this branch. The way you have done it violates that flow and is bound to cause some conflicts down the line.

I did branch off develop tho, I dont know why this is happening

**Comment by NicholusMuwonge** [comment]
> > > > Hi Nick, it appears that you are trying to merge two branches from this PR since you are merging two commits, one for this feature and another for the signup validation feature. Perhaps you accidentally created this branch off the signup validation branch, instead of from develop. Kindly fix this.
> > > 
> > > 
> > > Its because, I have two commits that are mine that need to be merged to develop at the same time
> > 
> > 
> > That's ok, I understand that you want to merge two commits however the convention is that each PR should merge 1 commit, which is my you have a PR for validating signup. Once the signup validation is approved and merged to develop, then you can rebase this branch with develop to add that feature to this branch. The way you have done it violates that flow and is bound to cause some conflicts down the line.
> 
> I did branch off develop tho, but let me try and fix it

**Review by reifred — APPROVED** [review]

**Review by reifred — CHANGES_REQUESTED** [review]
@NicholusMuwonge thanks for the good work done so far. However, when testing your feature as instructed on updating I don't get expected results. The bio and image aren't updated as expected.
```
PUT /api/user/
request body
{
	"user":{
		"username":"reifred",
    	"email": "reifred33@gmail.com",
    	"bio":"I work at a statefarm i",
    	"image": "https://inspector"
	}
}
Results.
{
    "user": {
        "email": "reifred33@gmail.com",
        "username": "reifred",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6NTIsInVzbiI6InJlaWZyZWQiLCJleHAiOjE1NTcxMzQzNzh9.7NlXTK6YREBIn9BAngC613zA38QoqZmm7FzUdAy7FjY",
        "bio": "",
        "image": ""
    }
}
```

**Comment by NicholusMuwonge** [comment]
> @NicholusMuwonge thanks for the good work done so far. However, when testing your feature as instructed on updating I don't get expected results. The bio and image aren't updated as expected.
> 
> ```
> PUT /api/user/
> request body
> {
> 	"user":{
> 		"username":"reifred",
>     	"email": "reifred33@gmail.com",
>     	"bio":"I work at a statefarm i",
>     	"image": "https://inspector"
> 	}
> }
> Results.
> {
>     "user": {
>         "email": "reifred33@gmail.com",
>         "username": "reifred",
>         "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6NTIsInVzbiI6InJlaWZyZWQiLCJleHAiOjE1NTcxMzQzNzh9.7NlXTK6YREBIn9BAngC613zA38QoqZmm7FzUdAy7FjY",
>         "bio": "",
>         "image": ""
>     }
> }
> ```

mine works fine, its probably the way u did your migrations or u might haave restored the removed content in the authentication/renderers.py if thats y this isnt working. I also added fix for image link so pull again and c the changes

**Comment by richien** [comment]
Hi @NicholusMuwonge, thank you for implementing the changes I requested. The commit looks ok now. However I have noticed that your PR is missing some sections, particularly the link to the PT story. Also note that there is an error in the PT storyID for this PR in the PR title as well as in the branch name; the correct storyID is [164691220](https://www.pivotaltracker.com/story/show/164691220) and not `16469122012` as in `ft-user-profile-16469122012` and `#16469122012 Add User profiles Functionality`

**Comment by reifred** [comment]
> > @NicholusMuwonge thanks for the good work done so far. However, when testing your feature as instructed on updating I don't get expected results. The bio and image aren't updated as expected.
> > ```
> > PUT /api/user/
> > request body
> > {
> > 	"user":{
> > 		"username":"reifred",
> >     	"email": "reifred33@gmail.com",
> >     	"bio":"I work at a statefarm i",
> >     	"image": "https://inspector"
> > 	}
> > }
> > Results.
> > {
> >     "user": {
> >         "email": "reifred33@gmail.com",
> >         "username": "reifred",
> >         "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6NTIsInVzbiI6InJlaWZyZWQiLCJleHAiOjE1NTcxMzQzNzh9.7NlXTK6YREBIn9BAngC613zA38QoqZmm7FzUdAy7FjY",
> >         "bio": "",
> >         "image": ""
> >     }
> > }
> > ```
> 
> mine works fine, its probably the way u did your migrations or u might haave restored the removed content in the authentication/renderers.py if thats y this isnt working. I also added fix for image link so pull again and c the changes

Thanks @NicholusMuwonge, for the updated work. Now it works fine. Work on the conflicting files for now though.

**Comment by NicholusMuwonge** [comment]
> Hi @NicholusMuwonge, thank you for implementing the changes I requested. The commit looks ok now. However I have noticed that your PR is missing some sections, particularly the link to the PT story. Also note that there is an error in the PT storyID for this PR in the PR title as well as in the branch name; the correct storyID is [164691220](https://www.pivotaltracker.com/story/show/164691220) and not `16469122012` as in `ft-user-profile-16469122012` and `#16469122012 Add User profiles Functionality`

Oh sorry please, Let me fix it

**Pull request closed without merging**